### PR TITLE
fixed #11339 : [v3.6] Crash after hot update by invoking cc.game.restart().

### LINKED ISSUE #11339: [v3.6] Crash after hot update
It's dangerous to use a thread that is detached.

WeChatd045b816d0efec0f19686a16d4572108
WeChatb2483f91871ca867ebd70bf33ae4e764

WeChatc9d23b370d6c5540a407838ee7b08dd2

WeChat456acf2d194dda4563e9be21df6715e2

## Step of code execute flow, read the `// -->` comments

```c++
void MessageQueue::runConsumerThread() noexcept {
    if (_immediateMode || _workerAttached) return;

    _reader.terminateConsumerThread = false;
    _reader.flushingFinished = false;

    std::thread consumerThread(&MessageQueue::consumerThreadLoop, this);
    consumerThread.detach(); // --> 1. consumerThreadLoop is a detached thread
    _workerAttached = true;
}

void DeviceAgent::doDestroy() {
    ......

    // --> 2. terminateConsumerThread in main thread
    _mainMessageQueue->terminateConsumerThread();

    _mainMessageQueue->~MessageQueue();
    CC_FREE_ALIGN(_mainMessageQueue);// --> 9. Delete MessageQueue instance, release its memory
    _mainMessageQueue = nullptr;
}

void MessageQueue::terminateConsumerThread() noexcept {
    if (_immediateMode || !_workerAttached) return;

    EventSem event;
    EventSem *const pEvent = &event;

    ReaderContext *const pR = &_reader;

    ccnew_placement(allocate<TerminateConsumerThreadMessage>(1)) TerminateConsumerThreadMessage(pEvent, pR); // --> 3. Send a terminate message to queue

    kick(); // --> 4. Tell the consumer thread to work
    event.wait(); // --> 5. Wait the last TerminateConsumerThreadMessage to be executed
    // --> 8. terminateConsumerThread done
}

void MessageQueue::consumerThreadLoop() noexcept {
    while (!_reader.terminateConsumerThread) {// --> 10. Bang~~~, MessageQueue was freed in step 9. Here checks the freed memory. Crash is found by Clang Address Sanitizer
        AutoReleasePool autoReleasePool;
        flushMessages(); // --> 6. flushMessages all messages in queue, handle the last terminate message
        // --> 8. continue to while loop
    }

    _workerAttached = false;// --> 11. Bang~~~, Here checks the freed memory second time. Crash is found by Clang Address Sanitizer
}

void TerminateConsumerThreadMessage::execute() noexcept {
    _reader->terminateConsumerThread = true;
    _reader->flushingFinished = true;
    _event->signal(); // --> 7. Notify main thread the last message was done, it just set two flags in reader.
}

```

### PR DESCRIPTION
**Pull request by dumganhar**
MessageQueue::terminateConsumerThread needs to wait consumer thread to exit. So don't detach consumer thread and we need to join the thread at the end of terminateConsumerThread.

Re: #11339

### Changelog

* Fix a crash after hot update by invoking cc.game.restart().

-------

### Continuous Integration

This pull request:

* [X] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/base/threading/MessageQueue.cpp [hunk 1]
```diff
@@ -141,6 +140,10 @@ void MessageQueue::terminateConsumerThread() noexcept {
 
     kick();
     event.wait();
+
+    if (_consumerThread->joinable()) {
+        _consumerThread->join();
```
Comment (star-e) [inline native/cocos/base/threading/MessageQueue.cpp]: after join, does it need to delete _consumerThread？

Comment (dumganhar) [reply native/cocos/base/threading/MessageQueue.cpp]: Yes, need delete it. I will fix it. Thank you.

**Review by star-e — APPROVED** [review]

**Comment by hana-alice** [comment]
what if we are running on triple-buffer? eg. MainmessageQueue can be running ahead of RenderQueue 3 frames before, now we have to wait until RenderQueue finished, which means it's almost running synchronously, in the aspect of frame.

**Comment by dumganhar** [comment]
> what if we are running on triple-buffer? eg. MainmessageQueue can be running ahead of RenderQueue 3 frames before, now we have to wait until RenderQueue finished, which means it's almost running synchronously, in the aspect of frame.

The key point of this issue is that the MessageQueue is deleted before RenderThread exits. So whether to use triple-buffer may not solve this issue. :(

**Comment by hana-alice** [comment]
yeah.
I mean does this pr has any effect on further implementation on triple-buffer? because we have to wait at end of the logic thread every frame, this may stops us from doing things like logic thread goes ahead of render thread 2 or 3 frames. And maybe this is why it's originally use 'detach'.

**Comment by dumganhar** [comment]
> I mean does this pr has any effect on further implementation on triple-buffer? because we have to wait at end of the logic thread every frame, this may stops us from doing things like logic thread goes ahead of render thread 2 or 3 frames. And maybe this is why it's originally use 'detach'.

@hana-alice  My change only affects the logic of destroying MessageQueue, not affects the logic of every frame.

**Review by hana-alice — APPROVED** [review]

File: native/cocos/base/threading/MessageQueue.cpp [hunk 2]
```diff
@@ -141,6 +140,13 @@ void MessageQueue::terminateConsumerThread() noexcept {
 
     kick();
     event.wait();
+
+    if (_consumerThread != nullptr) {
```
Comment (minggo) [inline native/cocos/base/threading/MessageQueue.cpp:144]: After `event.wait()`, the thread is not exited?

**Review by minggo — APPROVED** [review]

**Comment by xjtulk** [comment]
I met a similar crash about MessageQueue and i use v3.5.2 engine, After apply the same code change in my branch, it seems the crash disappear, but i am confused about the change, i think `event.Wait()` already ensure the thread exit cause it only be signaled when the TerminateConsumerThreadMessage is executed and the quit flags be set, so can you guys explain why the code works @dumganhar @minggo @hana-alice

**Comment by dumganhar** [comment]
@xjtulk The explanation is in the issue: https://github.com/cocos/cocos-engine/issues/11339

**Comment by xjtulk** [comment]
Thanks very much for the detailed explanation @dumganhar
